typed-list

esc opened the pull request:
Initial implementation of the typed list. Still quite a lot missing, but opening this PR to get an initial sanity check from @sklam

sklam commented:
I have read through the code.  It looks great!

ehsantn commented:
A side note: I have a piece of code that unboxes `Array[List[str]]` into Numba's `List[List[str]]` using the old implementation. I hope it's possible to port it to the new list implementation easily.

esc commented:
This is the pull-request for the Numba typed list implementation.

### What is included?

Much of the typed-list is now implemented. They are grouped into the following
pairs:

* `_listobject.{h,c}` - `tests/test_listimpl.py`        - the C level
* `listobject.py` - `tests/test_listobject.py`          - the compiler level
* `typed/typedlist.py` - `tests/test_typedlist.py`      - the interpreter level

And a few changes to ancillary files to enable compilation etc.. Comments can
be found at the top of the files to describe the approach taken and what the
code actually does.

### What is missing?

The following things are known to be issues:

- [ ] Using a slice in assignment to insert elements `a[i:j] = [x, y, z]` (slow)
- [ ] Using a slice in assignment to remove elements `a[i:j] = []` (slow)
- [ ] Delete a slice `del a[:]` (missing)
- [ ] Using a garbage array during deletion (missing / may not be needed)

The four items above are all related to mass deletion of objects, or mass
moving of objects. My current plan is to implement two c functions:
`numba_list_delete_slice(list, start, stop, step)` and
`numba_shift_left(list, from, to)`. The first will allow mass removal of
slices including a fast path for continuous slices (step=1/-1). The second will
allow making room for inserting items in the middle of the list somewhere. This
will probably also allow less code in `pop` at the C level and `insert` at the
compiler level.

- [ ] `__repr__` and `__str__` (missing)
- [ ] `index` (missing)
- [ ] `sort` (missing)
- [ ] Extensive checking for list mutation during iteration (missing)
- [ ] Equals method inside the method table (may not be needed)
- [ ] Extensive testing of the interpreter component (typed/typedlist.py)
- [ ] Inference Wiring for `list()` and `[]` (missing, dict was #4010)
- [ ] Definition of a minimal list size. (missing)
- [ ] Tests for when the list contains references to itself. (missing)
- [ ] Expose `item_size` and `allocated` from the c-level (missing)
- [ ] The windows `STACK_ALLOC` stuff (missing)

### What to look out for during review?

Since the code has been largely copied and modified from the typed-dict, there
may still be references to this. For example use of variable names `d`,
`dstruct`, `key`, `item` etc... Also, you will want to take a closer look at
all the stuff concerning the builder API, since this is what I am most
unfamiliar with. Also, at the C-level I am using `aligned_size` during
`malloc`. This was inherited from the typed-list and I am not sure if it is
really required. There are also still quite a few `FIXME` items that might need some
input.

Beyond that, please do try to break the implementation and find weird corner
cases that are not covered. The test-suites are pretty reasonable so suggesting
additional tests to expose bugs is more than welcome.

esc commented:
CI failed because an expected failure caused a memory leak, fixed that in a commit and pushed it.

esc commented:
I managed to fix the Python 2.7 compatibility issues and have pushed a fix.

esc commented:
Two critical issues with `resize` and `pop` that were detected when running under Linux have been fixed (hopefully) and pushed.

esc commented:
So, the Azure tests now all pass 🥁 🥁 🥁   and the code is really ready for the first more detailed review pass. As discussed with @stuartarchibald via google hangouts this morning, any additional changes will be submitted as additional pull-requests targetting this pull-request in order to keep the code being reviewed stable.

esc commented on numba/_listobject.c:
```diff
@@ -0,0 +1,522 @@
+valid_index(Py_ssize_t i, Py_ssize_t limit)
+{
+    /* The cast to size_t lets us use just a single comparison
+       to check whether i is in the range: 0 <= i < limit.
```
Code and comment were copied verbatim from: https://github.com/python/cpython/blob/master/Objects/listobject.c#L230

esc replied:
I thought about this some more and came to the conclusion that, if limit is 0, then we have an empty list and in that case 0 isn't a valid index.

esc commented on numba/_listobject.c:
```diff
@@ -0,0 +1,522 @@
+    // set method table to zero */
+    memset(&lp->methods, 0x00, sizeof(list_type_based_methods_table));
+    // allocate memory to hold items, if requested
+    if (allocated != 0) {
```
no, but it could be, if we decide to equip our list with a minimal size.

esc replied:
or, for example to accelerate things like `copy` because we can allocate a correctly sized list immediately, rather than growing it during the copy.

esc commented on numba/_listobject.c:
```diff
@@ -0,0 +1,522 @@
+    // allocate memory to hold items, if requested
+    if (allocated != 0) {
+        items = malloc(aligned_size(lp->item_size * allocated));
+        // allocated was definitely not zero, if malloc returns NULL
```
The only reason this is in there, is because I stumbled over this and it took me ages to debug.

esc replied:
I've documented the list struct some more in 7478ccf86f3ccb106bdc418cf6f43db3fe6bc101 including the 'allocated' member.

esc commented on numba/_listobject.c:
```diff
@@ -0,0 +1,522 @@
+        // pop from somewhere other than the end, incur the dreaded memory copy
+        leftover_bytes = (lp->size - 1 - index) * lp->item_size;
+        new_loc = lp->items + (lp->item_size * (index + 1));
+        // use memmove instead of memcpy, memcpy isn't safe on Linux
```
fixed in 7478ccf86f3ccb106bdc418cf6f43db3fe6bc101

esc commented on numba/_listobject.c:
```diff
@@ -0,0 +1,522 @@
+} ListStatus;
+
+static void
+copy_item(NB_List *lp, char *dst, const char *src){
```
Fixed in 7478ccf86f3ccb106bdc418cf6f43db3fe6bc101

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+    """
+    LIST_OK = 0,
+    LIST_ERR_INDEX = -1
+    LIST_ERR_NO_MEMORY = -2
```
Yes, should be fixed.

esc replied:
fixed in cef97cf5873841bb6a70f1e2d5c9dc19109fbc9b

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+    """Raise an internal error depending on the value of *status*
+    """
+    ok_status = status.type(int(ListStatus.LIST_OK))
+    with builder.if_then(builder.icmp_signed('!=', status, ok_status)):
```
Fixed in 0556a10776a295ea7f2ee686cb24ef5c57a45d70

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+
+
+@intrinsic
+def _from_meminfo(typingctx, mi, listtyperef):
```
Both `listobject.py` and `dictobject.py` live in the `numba` package. So I will place the shared code in a module `typedobjectutils.py` for now. Feel free to suggest an alternative name if you have a better one.

esc replied:
I had a look at the `_from` and `_as_meminfo` and they seem to be sufficiently different that they are not too amenable to generalization w/o some effort. I'll skip these for now.

esc replied:
And probably pick it up again for typed-set.

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+    return sig, codegen
+
+
+def _list_get_data(context, builder, list_ty, l):
```
Yes, this will work.

esc replied:
Fixed in bc0cf45912c81e1a5012d5f3714531f8870de296

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+
+
+# FIXME: copied from dictobject.py with minimal changes
+def _get_incref_decref(context, module, datamodel):
```
Yes, the comment alludes to it already.

esc replied:
Fixed in bc0cf45912c81e1a5012d5f3714531f8870de296

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+
+
+# FIXME: copied from dictobject.py with minimal changes
+def _get_equal(context, module, datamodel):
```
Yes.

esc replied:
Fixed in bc0cf45912c81e1a5012d5f3714531f8870de296

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+
+
+@intrinsic
+def _list_set_method_table(typingctx, lp, itemty):
```
The two methods tables are sufficiently different, key/value for dict and item for list, so this will probably not be that trivial.

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+
+
+@intrinsic
+def _make_list(typingctx, itemty, ptr):
```
Again, the differences between dict and list mean, that these two will be / are quite different.

stuartarchibald commented:
@esc the start of some additional tests to look at https://github.com/stuartarchibald/numba/commit/1bf503133a1a7ed7d1288f5c2d263c920a2168d8

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+def impl_len(l):
+    """len(list)
+    """
+    if not isinstance(l, types.ListType):
```
fixed in 1a0382ecc067a6626eaf07ae2a12d0524b75f59b

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+            if status == ListStatus.LIST_OK:
+                pass
+            elif status == ListStatus.LIST_ERR_INDEX:
+                raise IndexError("list index out of range from c")
```
The `IndexError` should have been raised above in `handle_index`, so this should never be reached. I will change this so that anything other than `LIST_OK` at this stage will lead to an `AssertionError`. The same pattern exists above in the `getitem` overload, so will fix there too.

esc replied:
Fixed in 275ed9a1a487ef41ab0f8a32b9400c3e380ea332

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+            casteditem = _cast(item, itemty)
+            status = _list_setitem(l, castedindex, casteditem)
+            if status == ListStatus.LIST_OK:
+                pass
```
Fixed in 275ed9a1a487ef41ab0f8a32b9400c3e380ea332

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+
+    elif isinstance(index, types.SliceType):
+        if not isinstance(item, types.IterableType):
+            raise TypingError("can only assign an iterable")
```
This faithfully reproduces the error message as in cpython:

```
In [36]: l[:] = 1
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-36-1d900bc915d0> in <module>
----> 1 l[:] = 1

TypeError: can only assign an iterable
```

None the less, I agree that this isn't exactly helpful, so I will come up with something more suitable.

esc replied:
Fixed in 48db76c44aa5a4d6052b855fdbd69ccfc53f0fd0

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+        def impl_slice(l, index, item):
+            # special case "a[i:j] = a", need to copy first
+            if l == item:
+                item = item.copy()
```
This is the copy from the overload, it does a shallow copy.

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+                    for i,j in zip(replace_range, item):
+                        l[i] = j
+                    # pop the remaining ones
+                    pop_range = range(slice_range.start + len(item),
```
This code has changed in https://github.com/esc/numba/pull/1

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+                    for i,j in zip(slice_range, item[:len(slice_range)]):
+                        l[i] = j
+                    # insert the remaining ones
+                    insert_range = range(slice_range.stop,
```
Maybe yes, but I am planning to replace this with something in C that will help with the expensive insert-in-a-loop below. I will revisit this then.

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+            if status == ListStatus.LIST_OK:
+                return _nonoptional(item)
+            elif status == ListStatus.LIST_ERR_INDEX:
+                raise IndexError("list index out of range")
```
yes there should be, it has a different  error message in cpython too.

esc replied:
Fixed in 08f825c2afc02e0e60108cc2a1916e6cf2187837

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+        return integer_impl
+
+    elif isinstance(index, types.SliceType):
+        def slice_impl(l, index):
```
Fixed in https://github.com/esc/numba/pull/1

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+
+
+@overload_method(types.ListType, 'insert')
+def impl_insert(l, index, item):
```
Yes, and they should use the `if index in types.signed_domain:` if feasible.

esc replied:
Fixed in e54096efb69694758374050af9c49373f3cab5cd

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+        # If the index is larger than the size of the list or if the list is
+        # empty, just append.
+        if index >= len(l) or len(l) == 0:
+            l.append(item)
```
`setitem`/ `append` should handle that.

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+
+    itemty = l.item_type
+
+    def impl(l):
```
Fixed in 0ddd2983964a41f43675a07b184bc3c8f15964eb

esc commented on numba/tests/test_listimpl.py:
```diff
@@ -0,0 +1,352 @@
+        ----------
+        tc : TestCase instance
+        item_size : int
+            byte size for the items
```
Fixed in 5b0a717681031d283d04de4abae5bf88168593f8

esc commented:
@stuartarchibald thanks for the review! For all the comments I have either responded or fixed the error and listed the commit hash of the fix.

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+def _list_getitem_pop_helper(typingctx, l, index, op):
+    """Wrap numba_list_getitem and numba_list_pop
+
+    Returns 2-tuple of (intp, ?item_type)
```
Yes, I copied this from the dict implementation without knowing exactly what it does. Probably worth removing this.

esc replied:
I had a look at removing this, but I think I will need some help. The use of the optional in this case allows us to propagate all the way to the overload and then raise the error there. Otherwise, you have to raise from the intrinsic.

esc commented on numba/_listobject.c:
```diff
@@ -0,0 +1,522 @@
+
+int
+numba_test_list(void) {
+    NB_List *lp;
```
Fixed in 7478ccf86f3ccb106bdc418cf6f43db3fe6bc101

esc commented on numba/_listobject.c:
```diff
@@ -0,0 +1,522 @@
+    CHECK(lp->size == 0);
+    CHECK(lp->allocated == 0);
+
+    // insert 1st item, this will cause a realloc
```
fixed in 7478ccf86f3ccb106bdc418cf6f43db3fe6bc101

esc commented on numba/_listobject.c:
```diff
@@ -0,0 +1,522 @@
+    CHECK(lp->allocated == 0);
+
+    // first, grow the list
+    // Use exactly 16 elements, should go through the allocation pattern:
```
Fixed in 7478ccf86f3ccb106bdc418cf6f43db3fe6bc101

esc commented on numba/_listobject.h:
```diff
@@ -0,0 +1,88 @@
+    /* Size of the list items. */
+    Py_ssize_t      item_size;
+
+    /* items contains space for 'allocated' elements.  The number
```
Fixed in 7478ccf86f3ccb106bdc418cf6f43db3fe6bc101

esc commented:
`index` is now implemented in https://github.com/esc/numba/pull/2

sklam commented on numba/listobject.py:
```diff
@@ -0,0 +1,1062 @@
+
+        dm_item = context.data_model_manager[itemty.instance_type]
+        if dm_item.contains_nrt_meminfo():
+            equal = _get_equal(context, builder.module, dm_item, 'list')
```
This will over-restrict the itemty to require the ability to do `==` on it.  This is currently preventing the use of jitclasses, which doesn't support `__eq__`.

esc replied:
The equals method in the method table isn't being used anywhere in the C level code and is more of an artifact left over from the copy and edit from the Dict. In could be removed in principle.

esc replied:
see https://github.com/esc/numba/pull/5

sklam commented on numba/listobject.py:
```diff
@@ -0,0 +1,1062 @@
+            for i in range(len(this)):
+                this_item, other_item = this[i], other[i]
+                # Cast item from LHS to the key-type of RHS
+                this_item = _cast(this_item, otheritemty)
```
This cast is unnecessary.

esc replied:
fixed in 86a1963d13458dfe59ef6d935fa30a643cfe69a0

sklam commented on numba/listobject.py:
```diff
@@ -0,0 +1,1062 @@
+            for i in range(len(this)):
+                this_item, other_item = this[i], other[i]
+                # Cast item from LHS to the key-type of RHS
+                this_item = _cast(this_item, otheritemty)
```
This cast is unnecessary.

esc replied:
fixed in 86a1963d13458dfe59ef6d935fa30a643cfe69a0

sklam commented:
@esc 
See my notebook for additional test cases: https://gist.github.com/sklam/b4d39561ff9afc9403222409e5696e66

- Ragged (awkward) list works. No action required.
- Jitclass in list doesn't work due to https://github.com/numba/numba/pull/4151#discussion_r298328592.
- Cyclic/recursive list tests may need inferred `.append()`. No action required.
- Reflectedlist in typedlist is broken because reflected list is buggy.  No action required.
- Typedlist in reflectedlist works. No action required.
- Misc:
  - Missing ordered comparison (i.e. `<`, `<=`, `>`, `>=`).  Good to have in this PR.
  - Missing `+` and `*`.  Good to have but lower priority IMO.

esc commented:
There was a pretty bad bug in the refinement of the list on `append`, I had to modify some critical code in `numba/typing/templates.py` so this should be thoroughly checked:

https://github.com/esc/numba/pull/4/files

esc commented:
This removed the method table, and should allow use of non-comparable classes like jitclass:

https://github.com/esc/numba/pull/5

esc commented on numba/listobject.py:
```diff
@@ -0,0 +1,1089 @@
+
+    itemty = l.item_type
+
+    def impl(l, iterable):
```
This is fixed in 7ca2ecb92711fab9bae20f04f691bc50fecc7b53

esc commented:
Comparison operators are in https://github.com/esc/numba/pull/6 - albeit with memory leaks..

esc commented:
I will merge in current master, there are a bunch of fixes there now that we will need for this.

esc commented:
So, comparison operators, inference of `List()` in jit context and storing `jitclass`es should now work.

sklam commented:
Looks like only `.append` has an effect to type-inferred list usage.   Other mutating methods should either work or produce an error.  

- `.extend`
- `.insert`

esc commented:
@sklam type inference for `insert` is fixed in 36f187831bcf133a7dafdfaced1900a1274cb07c

esc commented:
@sklam type refinement for extend is in  f543ea8edeec68ff7f03475913dda5fe6268eb1d - but with limitations.

sklam commented:
`.insert` and `.extend` looks good

stuartarchibald commented on numba/_listobject.h:
```diff
@@ -0,0 +1,99 @@
+/* Adapted from CPython3.7 Include/listobject.h */
```
This needs a reference URL with an exact SHA from the `python/cpython` github repo: https://github.com/python/cpython

esc replied:
fixed with fbd34400b684ca7ad67f3e547a28a5e40a4ffd16

esc commented:
Alrighty, docs have been merged in now too.

esc commented:
Pending build farm and 32 bit builds.

seibert commented:
I think everything has passed on the buildfarm we needed to check.  Marking for merge!

esc commented:
🎉 🎉 🎉
